### PR TITLE
use hash-base

### PR DESCRIPTION
**Pull request by fanatid**
Ref #25

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,53 +1,40 @@
...
-
-## Acknowledgements
-
-This work is derived from Paul Johnston's ["A JavaScript implementation of the Secure Hash Algorithm"]
-(http://pajhome.org.uk/crypt/md5/sha1.html)
-
+`sha.js` currently implements:
 
+  - SHA (SHA-0) -- **legacy, no not use in new systems**
+  - SHA-1 -- **legacy, no not use in new systems**
```
Comment (dcousens) [inline README.md]: **do** not,  not **no** not

Comment (fanatid) [inline README.md]: Fixed, thanks.
I doubted about this phrase, but decided not change it when edited README.md

File: lib/sha1.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+'use strict'
+var inherits = require('inherits')
+var SHA = require('./sha')
+
+var ARRAY80 = new Array(80)
```
Comment (dcousens) [inline lib/sha1.js]: Maybe just call this `var W = new Array(80)`,  and let it be super clear that it is a global used by `_getExpandedMessage`

File: lib/sha.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+SHA.prototype.__digest = function () {
+  var buffer = new Buffer(20)
+  buffer.writeInt32BE(this._a, 0)
+  buffer.writeInt32BE(this._b, 4)
+  buffer.writeInt32BE(this._c, 8)
+  buffer.writeInt32BE(this._d, 12)
+  buffer.writeInt32BE(this._e, 16)
+  return buffer
+}
+
```
Comment (dcousens) [inline lib/sha.js]: Can we opt for functions at the top?  Not bottom?

File: lib/sha256.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  buffer.writeInt32BE(this._b, 4)
+  buffer.writeInt32BE(this._c, 8)
+  buffer.writeInt32BE(this._d, 12)
+  buffer.writeInt32BE(this._e, 16)
+  buffer.writeInt32BE(this._f, 20)
+  buffer.writeInt32BE(this._g, 24)
+  buffer.writeInt32BE(this._h, 28)
+  return buffer
+}
+
```
Comment (dcousens) [inline lib/sha256.js]: Same here,  functions *before* use?

File: lib/sha512.js [hunk 5]
```diff
@@ -44,147 +46,104 @@ var K = [
...
-  var bl = this._bl | 0
-  var cl = this._cl | 0
-  var dl = this._dl | 0
-  var el = this._el | 0
-  var fl = this._fl | 0
-  var gl = this._gl | 0
-  var hl = this._hl | 0
-
+SHA512.prototype._update = function (M) {
+  var W = ARRAY160
```
Comment (dcousens) [inline lib/sha512.js]: Same here,  don't bother with the global var `ARRAY160`,  just be explicit with calling it `W` to start?

File: lib/sha512.js [hunk 6]
```diff
@@ -44,147 +46,104 @@ var K = [
...
   for (var i = 0; i < 32; i += 2) {
-    W[i] = M.readInt32BE(i * 4)
-    W[i + 1] = M.readInt32BE(i * 4 + 4)
+    W[i] = this._block.readInt32BE(i * 4)
+    W[i + 1] = this._block.readInt32BE(i * 4 + 4)
   }
-  for (; i < 160; i += 2) {
+  for (i = 32; i < 160; i += 2) {
```
Comment (dcousens) [inline lib/sha512.js]: Is the `i = 32` necessary?  Not that it matters.

File: lib/sha256.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  this._c = 0x3c6ef372
+  this._d = 0xa54ff53a
+  this._e = 0x510e527f
+  this._f = 0x9b05688c
+  this._g = 0x1f83d9ab
+  this._h = 0x5be0cd19
+}
+
+Sha256.prototype._update = function () {
+  var W = ARRAY64
```
Comment (dcousens) [inline lib/sha256.js]: And here

File: lib/sha256.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  for (var i = 0; i < 16; ++i) W[i] = this._block.readInt32BE(i * 4)
+  for (; i < 64; ++i) W[i] = (gamma1(W[i - 2]) + W[i - 7] + gamma0(W[i - 15]) + W[i - 16]) | 0
```
Comment (dcousens) [inline lib/sha256.js]: No `i = 16` this time?

**Comment by dcousens** [comment]
utACK e6e054c,

**Comment by fanatid** [comment]
@dcousens updated

**Comment by dcousens** [comment]
Published as `+ sha.js@2.4.6`

**Comment by dcousens** [comment]
Why did you move the files to `lib/`?
Seems a bit unnecessary IMHO.

File: lib/sha.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  this._c = 0x98badcfe
+  this._d = 0x10325476
+  this._e = 0xc3d2e1f0
+}
+
+inherits(SHA, HashBase)
+
+SHA.prototype._expandMessage = function (W) {
+  for (var i = 0; i < 16; ++i) W[i] = this._block.readInt32BE(i * 4)
+  for (; i < 80; ++i) W[i] = W[i - 3] ^ W[i - 8] ^ W[i - 14] ^ W[i - 16]
```
Comment (dcousens) [inline lib/sha.js]: Where did this come from/

**Comment by dcousens** [comment]
Why did you remove `.bin`... its might be used by people.

**Comment by dcousens** [comment]
Re-added `bin.js`,  released as `v2.4.7`

**Comment by dcousens** [comment]
@fanatid please keep PRs atomic.

Don't remove things in the same PR as something as useful as solving #25.

File: index.js [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 
   return new Algorithm()
 }
 
-exports.sha = require('./sha')
-exports.sha1 = require('./sha1')
-exports.sha224 = require('./sha224')
-exports.sha256 = require('./sha256')
-exports.sha384 = require('./sha384')
-exports.sha512 = require('./sha512')
```
Comment (dcousens) [inline index.js]: Sigh

**Comment by dcousens** [comment]
@fanatid also please mark things as major bump if they are.

Unpublished `v2.4.7`.

**Comment by dcousens** [comment]
@fanatid can you please re-add the tests, un-change the exports to avoid breaking backwards compatibility and then submit that PR,  so we can release this as a non-breaking change?

**Comment by fanatid** [comment]
@dcousens please revert all change and unpublish, I can't add tests/refactor PR right now.
I mentioned in #39 that this PR need major bump.
Sorry for not atomic commits.

**Comment by dcousens** [comment]
@fanatid I already unpublished the problematic changes.
The change need not be reverted, it just needs to be fixed to be non-breaking :+1:

> I mentioned in #39 that this PR need major bump.

I didn't see that until after.

**Comment by fanatid** [comment]
I think hash-base makes breaking changes anyway.
utf8 instead binary by default for example

**Comment by dcousens** [comment]
@fanatid not willing to be backwards compatible in `hash-base`?

**Comment by fanatid** [comment]
are you think that this necessary? `0.12` was last version of node who uses binary by default

**Comment by gethinwebster** [comment]
Just to note, as others may come across this as well, this update broke some of our projects which use this module, but don't use browserify. `hash-base` pulls in a new dependency of `stream`, which wasn't previously required, and although it's provided as a shim by default in browserify, other packagers vary.

**Comment by fanatid** [comment]
@gethinwebster sorry hear this, but is not `2.4.7` fix `hash-base` changes?

**Comment by gethinwebster** [comment]
It doesn't seem to - lolks like 2.4.7 still uses hash-base unless I'm missing something?

**Comment by fanatid** [comment]
ah, I see :(
@dcousens what you think about publish `2.4.8` equal to old `2.4.6` and unpublish `2.4.7`?

**Comment by gethinwebster** [comment]
2.4.5 is the latest version that works correctly for me

**Comment by dominictarr** [comment]
Ah, so moving the files into lib was a breaking change.
I was using it like this:
https://github.com/dominictarr/sodium-browserify-tweetnacl/blob/master/index.js#L3

and so I have lots of stuff that is broken now. If it goes into lib it needs a major semver bump

**Comment by dcousens** [comment]
Agreed, I commonly use it like `sha.js/sha256` as well.
Will need to revert `lib/` as well.

- [ ] Revert `lib/`
- [ ] Revert export uppercasing
- [ ] Revert `bin.js` removal

**Comment by dcousens** [comment]
Unpublish wasn't working as expected for `v2.4.7`.
Reverted all changes and published as `v2.4.8`.

@dominictarr @fanatid please [verify](https://github.com/crypto-browserify/sha.js/commits/master).

**Comment by gethinwebster** [comment]
Thanks for the really quick response on this, great work @dcousens @fanatid

**Comment by dominictarr** [comment]
yup, working again.

**Comment by dcousens** [comment]
@fanatid please submit atomic PRs for the changes :+1:

**Comment by dcousens** [comment]
@fanatid what were the blocks here?
I think we could default to `utf8` over `binary` using the same trick we did in `pbkdf2`?

**Comment by larcho** [comment]
I know this is a very very VERY old PR, but I'd like to double check something. I actually do like this version because it uses `readable-stream` from hash-base instead of standard Browser/Node Stream. React Native does not include Stream.
I'm aware of the `rn-nodeify`, but I'm updating a library that depends on the `crypto` library to not require "nodeification".
Is there any reason I should not use this version?

**Pull request merged**